From 846c6e51e5d8b6c888ba448a9b25fdb62397ae2f Mon Sep 17 00:00:00 2001
From: Shashank Babu Chinta Venkata <schintav@qti.qualcomm.com>
Date: Thu, 26 Jun 2025 15:04:28 -0700
Subject: [PATCH] PENDING: firmware: extend vmid support to 128

Current implementation only supports source vms under 64
during mem assign API call. With vmid now extending upto 128,
the callers need more than single word to express
source vm in bitmap format.

Change-Id: If9748cc8ee78bf4c0e701b0bc85c171483ade26b
Signed-off-by: Shashank Babu Chinta Venkata <schintav@qti.qualcomm.com>
---
 .../heaps/qcom_dma_heap_secure_utils.c        |  5 ++-
 drivers/firmware/qcom_scm.c                   | 29 +++++++++----
 drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c    | 12 +++---
 drivers/misc/fastrpc.c                        | 42 +++++++++++--------
 drivers/net/wireless/ath/ath10k/qmi.c         | 14 +++----
 drivers/soc/qcom/rmtfs_mem.c                  | 10 ++---
 drivers/virt/gunyah/gunyah_qcom.c             | 23 +++++-----
 include/linux/firmware/qcom/qcom_scm.h        |  3 ++
 sound/soc/qcom/qdsp6/msm_audio_mem.c          | 15 ++++---
 9 files changed, 92 insertions(+), 61 deletions(-)

diff --git a/drivers/dma-buf/heaps/qcom_dma_heap_secure_utils.c b/drivers/dma-buf/heaps/qcom_dma_heap_secure_utils.c
index a50b0fed58a2..9763cecfdc72 100644
--- a/drivers/dma-buf/heaps/qcom_dma_heap_secure_utils.c
+++ b/drivers/dma-buf/heaps/qcom_dma_heap_secure_utils.c
@@ -256,11 +256,12 @@ int hyp_assign_from_flags(u64 base, u64 size, unsigned long flags)
 {
 	int *vmids, *modes;
 	u32 nr;
-	u64 src_vm = BIT(QCOM_SCM_VMID_HLOS);
+	u64 src_vm [2] = {0};
 	struct qcom_scm_vmperm *newvm;
 	int ret;
 	int i;
 
+	src_vm[0] = BIT(QCOM_SCM_VMID_HLOS);
 	ret = get_vmperm_from_ion_flags(flags, &vmids, &modes, &nr);
 	if (ret)
 		return ret;
@@ -274,7 +275,7 @@ int hyp_assign_from_flags(u64 base, u64 size, unsigned long flags)
 		newvm[i].vmid = vmids[i];
 		newvm[i].perm = modes[i];
 	}
-	ret = qcom_scm_assign_mem(base, size, &src_vm, newvm, nr);
+	ret = qcom_scm_assign_mem(base, size, &src_vm[0], newvm, nr);
 	if (ret)
 		pr_err("%s: Assign call failed, flags 0x%lx\n", __func__,
 		       flags);
diff --git a/drivers/firmware/qcom_scm.c b/drivers/firmware/qcom_scm.c
index 36edd96a84eb..b77c3ce1395a 100644
--- a/drivers/firmware/qcom_scm.c
+++ b/drivers/firmware/qcom_scm.c
@@ -1006,6 +1006,16 @@ static int __qcom_scm_assign_mem(struct device *dev, phys_addr_t mem_region,
 	return ret ? : res.result[0];
 }
 
+void qcom_scm_set_vmid_by_word(u64 *srcvm, int vmid)
+{
+	if (vmid < QCOM_SCM_VM_WORD_SIZE) {
+		srcvm[0] |= BIT(vmid);
+	} else {
+		srcvm[1] |= BIT(vmid - QCOM_SCM_VM_WORD_SIZE);
+	}
+}
+EXPORT_SYMBOL_GPL(qcom_scm_set_vmid_by_word);
+
 /**
  * qcom_scm_assign_mem() - Make a secure call to reassign memory ownership
  * @mem_addr: mem region whose ownership need to be reassigned
@@ -1032,12 +1042,11 @@ int qcom_scm_assign_mem(phys_addr_t mem_addr, size_t mem_sz,
 	size_t dest_sz;
 	size_t src_sz;
 	size_t ptr_sz;
-	int next_vm;
+	u64 next_vm[2] = {0};
 	__le32 *src;
 	int ret, i, b;
-	u64 srcvm_bits = *srcvm;
 
-	src_sz = hweight64(srcvm_bits) * sizeof(*src);
+	src_sz = (hweight64(srcvm[0]) + (hweight64(srcvm[1]))) * sizeof(*src);
 	mem_to_map_sz = sizeof(*mem_to_map);
 	dest_sz = dest_cnt * sizeof(*destvm);
 	ptr_sz = ALIGN(src_sz, SZ_64) + ALIGN(mem_to_map_sz, SZ_64) +
@@ -1054,7 +1063,12 @@ int qcom_scm_assign_mem(phys_addr_t mem_addr, size_t mem_sz,
 	src = ptr;
 	i = 0;
 	for (b = 0; b < BITS_PER_TYPE(u64); b++) {
-		if (srcvm_bits & BIT(b))
+		if (srcvm[0] & BIT(b))
+			src[i++] = cpu_to_le32(b);
+	}
+
+	for (b = QCOM_SCM_VM_WORD_SIZE; b < QCOM_SCM_MAX_VMID; b++) {
+		if (srcvm[1] & BIT(b - QCOM_SCM_VM_WORD_SIZE))
 			src[i++] = cpu_to_le32(b);
 	}
 
@@ -1064,7 +1078,6 @@ int qcom_scm_assign_mem(phys_addr_t mem_addr, size_t mem_sz,
 	mem_to_map->mem_addr = cpu_to_le64(mem_addr);
 	mem_to_map->mem_size = cpu_to_le64(mem_sz);
 
-	next_vm = 0;
 	/* Fill details of next vmid detail */
 	destvm = ptr + ALIGN(mem_to_map_sz, SZ_64) + ALIGN(src_sz, SZ_64);
 	dest_phys = ptr_phys + ALIGN(mem_to_map_sz, SZ_64) + ALIGN(src_sz, SZ_64);
@@ -1073,7 +1086,7 @@ int qcom_scm_assign_mem(phys_addr_t mem_addr, size_t mem_sz,
 		destvm->perm = cpu_to_le32(newvm->perm);
 		destvm->ctx = 0;
 		destvm->ctx_size = 0;
-		next_vm |= BIT(newvm->vmid);
+		qcom_scm_set_vmid_by_word(&next_vm[0], newvm->vmid);
 	}
 
 	ret = __qcom_scm_assign_mem(__scm->dev, mem_to_map_phys, mem_to_map_sz,
@@ -1081,10 +1094,12 @@ int qcom_scm_assign_mem(phys_addr_t mem_addr, size_t mem_sz,
 	if (ret) {
 		dev_err(__scm->dev,
 			"Assign memory protection call failed %d\n", ret);
+		WARN_ON(1);
 		return -EINVAL;
 	}
 
-	*srcvm = next_vm;
+	srcvm[0] = next_vm[0];
+	srcvm[1] = next_vm[1];
 	return 0;
 }
 EXPORT_SYMBOL_GPL(qcom_scm_assign_mem);
diff --git a/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c b/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
index 828c7488ee47..da4a22b9ad18 100644
--- a/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
+++ b/drivers/iommu/arm/arm-smmu/arm-smmu-qcom.c
@@ -422,11 +422,12 @@ static void *qcom_alloc_pages(void *cookie, size_t size, gfp_t gfp)
 {
 	struct page *p;
 	struct qcom_scm_vmperm perms[2];
-	u64 src  = BIT(QCOM_SCM_VMID_HLOS);
+	u64 src[2] = {0};
 	int ret;
 
 	struct arm_smmu_domain *domain = (void *)cookie;
 
+	src[0] = BIT(QCOM_SCM_VMID_HLOS);
 	p = alloc_page(gfp);
 	if (!p)
 		return NULL;
@@ -436,7 +437,7 @@ static void *qcom_alloc_pages(void *cookie, size_t size, gfp_t gfp)
 	perms[1].vmid = domain->secure_vmid;
 	perms[1].perm = QCOM_SCM_PERM_READ;
 	ret = qcom_scm_assign_mem(page_to_phys(p), PAGE_SIZE,
-				  &src, perms, 2);
+				  &src[0], perms, 2);
 	if (ret < 0) {
 		dev_err(domain->smmu->dev,
 			"assign memory failed for vmid=%x ret=%d\n",
@@ -452,7 +453,7 @@ static void qcom_free_pages(void *cookie, void *pages, size_t size)
 {
 	struct qcom_scm_vmperm perms;
 	struct page *p;
-	u64 src;
+	u64 src[2] = {0};
 	int ret;
 
 	struct arm_smmu_domain *domain = (void *)cookie;
@@ -461,9 +462,10 @@ static void qcom_free_pages(void *cookie, void *pages, size_t size)
 
 	perms.vmid = QCOM_SCM_VMID_HLOS;
 	perms.perm = QCOM_SCM_PERM_RWX;
-	src = BIT(domain->secure_vmid) | BIT(QCOM_SCM_VMID_HLOS);
+	src[0] |= BIT(QCOM_SCM_VMID_HLOS);
+	qcom_scm_set_vmid_by_word(&src[0], domain->secure_vmid);
 	ret = qcom_scm_assign_mem(page_to_phys(p), PAGE_SIZE,
-				  &src, &perms, 1);
+				  &src[0], &perms, 1);
 	/*
 	 * For assign failure scenario, it is not safe to use these pages by HLOS.
 	 * So returning from here instead of freeing the page.
diff --git a/drivers/misc/fastrpc.c b/drivers/misc/fastrpc.c
index 417d9f490cbb..131c4684943f 100644
--- a/drivers/misc/fastrpc.c
+++ b/drivers/misc/fastrpc.c
@@ -456,13 +456,14 @@ static void fastrpc_free_map(struct kref *ref)
 		if (map->attr & FASTRPC_ATTR_SECUREMAP) {
 			struct qcom_scm_vmperm perm;
 			int vmid = fl->cctx->vmperms[0].vmid;
-			u64 src_perms = BIT(QCOM_SCM_VMID_HLOS) | BIT(vmid);
 			int err = 0;
-
+			u64 src_perms[2] = {0};
+			src_perms[0] = BIT(QCOM_SCM_VMID_HLOS);
+			qcom_scm_set_vmid_by_word(&src_perms[0], vmid);
 			perm.vmid = QCOM_SCM_VMID_HLOS;
 			perm.perm = QCOM_SCM_PERM_RWX;
 			err = qcom_scm_assign_mem(map->phys, map->len,
-				&src_perms, &perm, 1);
+				&src_perms[0], &perm, 1);
 			if (err) {
 				dev_err(fl->sctx->dev, "Failed to assign memory phys 0x%llx size 0x%llx err %d\n",
 						map->phys, map->len, err);
@@ -992,15 +993,16 @@ static int fastrpc_map_create(struct fastrpc_user *fl, int fd,
 		 * If subsystem VMIDs are defined in DTSI, then do
 		 * hyp_assign from HLOS to those VM(s)
 		 */
-		u64 src_perms = BIT(QCOM_SCM_VMID_HLOS);
+		u64 src_perms[2] = {0};
 		struct qcom_scm_vmperm dst_perms[2] = {0};
 
+		src_perms[0] = BIT(QCOM_SCM_VMID_HLOS);
 		dst_perms[0].vmid = QCOM_SCM_VMID_HLOS;
 		dst_perms[0].perm = QCOM_SCM_PERM_RW;
 		dst_perms[1].vmid = fl->cctx->vmperms[0].vmid;
 		dst_perms[1].perm = QCOM_SCM_PERM_RWX;
 		map->attr = attr;
-		err = qcom_scm_assign_mem(map->phys, (u64)map->len, &src_perms, dst_perms, 2);
+		err = qcom_scm_assign_mem(map->phys, (u64)map->len, &src_perms[0], dst_perms, 2);
 		if (err) {
 			dev_err(dev, "Failed to assign memory with phys 0x%llx size 0x%llx err %d\n",
 					map->phys, map->len, err);
@@ -1589,17 +1591,17 @@ static int fastrpc_mmap_remove_ssr(struct fastrpc_channel_ctx *cctx)
 
 	list_for_each_entry_safe(buf, b, &cctx->rhmaps, node) {
 		if (cctx->vmcount) {
-			u64 src_perms = 0;
+			u64 src_perms[2] = {0};
 			struct qcom_scm_vmperm dst_perms;
 			u32 i;
 
 			for (i = 0; i < cctx->vmcount; i++)
-				src_perms |= BIT(cctx->vmperms[i].vmid);
+				qcom_scm_set_vmid_by_word(&src_perms[0], cctx->vmperms[i].vmid);
 
 			dst_perms.vmid = QCOM_SCM_VMID_HLOS;
 			dst_perms.perm = QCOM_SCM_PERM_RWX;
 			err = qcom_scm_assign_mem(buf->phys, (u64)buf->size,
-				&src_perms, &dst_perms, 1);
+				&src_perms[0], &dst_perms, 1);
 			if (err) {
 				pr_err("%s: Failed to assign memory phys 0x%llx size 0x%llx err %d\n",
 					__func__, buf->phys, buf->size, err);
@@ -1704,10 +1706,11 @@ static int fastrpc_init_create_static_process(struct fastrpc_user *fl,
 		size = buf->size;
 		/* Map if we have any heap VMIDs associated with this ADSP Static Process. */
 		if (fl->cctx->vmcount) {
-			u64 src_perms = BIT(QCOM_SCM_VMID_HLOS);
+			u64 src_perms[2] = {0};
+			src_perms[0] = BIT(QCOM_SCM_VMID_HLOS);
 
 			err = qcom_scm_assign_mem(phys, (u64)size,
-					&src_perms, fl->cctx->vmperms,
+					&src_perms[0], fl->cctx->vmperms,
 					fl->cctx->vmcount);
 			if (err) {
 				dev_err(fl->sctx->dev, "Failed to assign memory with phys 0x%llx size 0x%llx err %d",
@@ -1755,17 +1758,18 @@ static int fastrpc_init_create_static_process(struct fastrpc_user *fl,
 	return 0;
 err_invoke:
 	if (fl->cctx->vmcount && scm_done) {
-		u64 src_perms = 0;
+		u64 src_perms[2] = {0};
 		struct qcom_scm_vmperm dst_perms;
 		u32 i;
 
 		for (i = 0; i < fl->cctx->vmcount; i++)
-			src_perms |= BIT(fl->cctx->vmperms[i].vmid);
+			qcom_scm_set_vmid_by_word(&src_perms[0],
+							fl->cctx->vmperms[i].vmid);
 
 		dst_perms.vmid = QCOM_SCM_VMID_HLOS;
 		dst_perms.perm = QCOM_SCM_PERM_RWX;
 		err = qcom_scm_assign_mem(phys, (u64)size,
-						&src_perms, &dst_perms, 1);
+						&src_perms[0], &dst_perms, 1);
 		if (err)
 			dev_err(fl->sctx->dev, "Failed to assign memory phys 0x%llx size 0x%llx err %d",
 				phys, size, err);
@@ -2296,17 +2300,18 @@ static int fastrpc_req_munmap_impl(struct fastrpc_user *fl, struct fastrpc_buf *
 	if (!err) {
 		if (buf->flag == ADSP_MMAP_REMOTE_HEAP_ADDR) {
 			if (fl->cctx->vmcount) {
-				u64 src_perms = 0;
+				u64 src_perms[2] = {0};
 				struct qcom_scm_vmperm dst_perms;
 				u32 i;
 
 				for (i = 0; i < fl->cctx->vmcount; i++)
-					src_perms |= BIT(fl->cctx->vmperms[i].vmid);
+					qcom_scm_set_vmid_by_word(&src_perms[0],
+							fl->cctx->vmperms[i].vmid);
 
 				dst_perms.vmid = QCOM_SCM_VMID_HLOS;
 				dst_perms.perm = QCOM_SCM_PERM_RWX;
 				err = qcom_scm_assign_mem(buf->phys, (u64)buf->size,
-								&src_perms, &dst_perms, 1);
+								&src_perms[0], &dst_perms, 1);
 				if (err) {
 					dev_err(fl->sctx->dev, "Failed to assign memory phys 0x%llx size 0x%llx err %d\n",
 						buf->phys, buf->size, err);
@@ -2470,10 +2475,11 @@ static int fastrpc_req_mmap(struct fastrpc_user *fl, char __user *argp)
 
 		/* Add memory to static PD pool, protection thru hypervisor */
 		if (req.flags == ADSP_MMAP_REMOTE_HEAP_ADDR && fl->cctx->vmcount) {
-			u64 src_perms = BIT(QCOM_SCM_VMID_HLOS);
+			u64 src_perms[2] = {0};
+			src_perms[0] = BIT(QCOM_SCM_VMID_HLOS);
 
 			err = qcom_scm_assign_mem(buf->phys, (u64)buf->size,
-				&src_perms, fl->cctx->vmperms, fl->cctx->vmcount);
+				&src_perms[0], fl->cctx->vmperms, fl->cctx->vmcount);
 			if (err) {
 				dev_err(fl->sctx->dev, "Failed to assign memory phys 0x%llx size 0x%llx err %d\n",
 						buf->phys, buf->size, err);
diff --git a/drivers/net/wireless/ath/ath10k/qmi.c b/drivers/net/wireless/ath/ath10k/qmi.c
index 38e939f572a9..418961c3a57f 100644
--- a/drivers/net/wireless/ath/ath10k/qmi.c
+++ b/drivers/net/wireless/ath/ath10k/qmi.c
@@ -34,11 +34,11 @@ static int ath10k_qmi_map_msa_permission(struct ath10k_qmi *qmi,
 {
 	struct qcom_scm_vmperm dst_perms[3];
 	struct ath10k *ar = qmi->ar;
-	u64 src_perms;
+	u64 src_perms[2] = {0};
 	u32 perm_count;
 	int ret;
 
-	src_perms = BIT(QCOM_SCM_VMID_HLOS);
+	src_perms[0] = BIT(QCOM_SCM_VMID_HLOS);
 
 	dst_perms[0].vmid = QCOM_SCM_VMID_MSS_MSA;
 	dst_perms[0].perm = QCOM_SCM_PERM_RW;
@@ -54,7 +54,7 @@ static int ath10k_qmi_map_msa_permission(struct ath10k_qmi *qmi,
 	}
 
 	ret = qcom_scm_assign_mem(mem_info->addr, mem_info->size,
-				  &src_perms, dst_perms, perm_count);
+				  &src_perms[0], dst_perms, perm_count);
 	if (ret < 0)
 		ath10k_err(ar, "failed to assign msa map permissions: %d\n", ret);
 
@@ -66,19 +66,19 @@ static int ath10k_qmi_unmap_msa_permission(struct ath10k_qmi *qmi,
 {
 	struct qcom_scm_vmperm dst_perms;
 	struct ath10k *ar = qmi->ar;
-	u64 src_perms;
+	u64 src_perms[2] = {0};
 	int ret;
 
-	src_perms = BIT(QCOM_SCM_VMID_MSS_MSA) | BIT(QCOM_SCM_VMID_WLAN);
+	src_perms[0] = BIT(QCOM_SCM_VMID_MSS_MSA) | BIT(QCOM_SCM_VMID_WLAN);
 
 	if (!mem_info->secure)
-		src_perms |= BIT(QCOM_SCM_VMID_WLAN_CE);
+		src_perms[0] |= BIT(QCOM_SCM_VMID_WLAN_CE);
 
 	dst_perms.vmid = QCOM_SCM_VMID_HLOS;
 	dst_perms.perm = QCOM_SCM_PERM_RW;
 
 	ret = qcom_scm_assign_mem(mem_info->addr, mem_info->size,
-				  &src_perms, &dst_perms, 1);
+				  &src_perms[0], &dst_perms, 1);
 	if (ret < 0)
 		ath10k_err(ar, "failed to unmap msa permissions: %d\n", ret);
 
diff --git a/drivers/soc/qcom/rmtfs_mem.c b/drivers/soc/qcom/rmtfs_mem.c
index f83811f51175..403e5a3ba2f8 100644
--- a/drivers/soc/qcom/rmtfs_mem.c
+++ b/drivers/soc/qcom/rmtfs_mem.c
@@ -31,7 +31,7 @@ struct qcom_rmtfs_mem {
 
 	unsigned int client_id;
 
-	u64 perms;
+	u64 perms[2];
 };
 
 static ssize_t qcom_rmtfs_mem_show(struct device *dev,
@@ -260,9 +260,9 @@ static int qcom_rmtfs_mem_probe(struct platform_device *pdev)
 			perms[i + 1].perm = QCOM_SCM_PERM_RW;
 		}
 
-		rmtfs_mem->perms = BIT(QCOM_SCM_VMID_HLOS);
+		rmtfs_mem->perms[0] = BIT(QCOM_SCM_VMID_HLOS);
 		ret = qcom_scm_assign_mem(rmtfs_mem->addr, rmtfs_mem->size,
-					  &rmtfs_mem->perms, perms, num_vmids + 1);
+					  &rmtfs_mem->perms[0], perms, num_vmids + 1);
 		if (ret < 0) {
 			dev_err(&pdev->dev, "assign memory failed\n");
 			goto remove_cdev;
@@ -286,12 +286,12 @@ static int qcom_rmtfs_mem_remove(struct platform_device *pdev)
 	struct qcom_rmtfs_mem *rmtfs_mem = dev_get_drvdata(&pdev->dev);
 	struct qcom_scm_vmperm perm;
 
-	if (rmtfs_mem->perms) {
+	if (rmtfs_mem->perms[0]) {
 		perm.vmid = QCOM_SCM_VMID_HLOS;
 		perm.perm = QCOM_SCM_PERM_RW;
 
 		qcom_scm_assign_mem(rmtfs_mem->addr, rmtfs_mem->size,
-				    &rmtfs_mem->perms, &perm, 1);
+				    &rmtfs_mem->perms[0], &perm, 1);
 	}
 
 	cdev_device_del(&rmtfs_mem->cdev, &rmtfs_mem->dev);
diff --git a/drivers/virt/gunyah/gunyah_qcom.c b/drivers/virt/gunyah/gunyah_qcom.c
index f06a598f2e1b..4a7df3c46842 100644
--- a/drivers/virt/gunyah/gunyah_qcom.c
+++ b/drivers/virt/gunyah/gunyah_qcom.c
@@ -16,7 +16,8 @@
 static int qcom_scm_gh_rm_pre_mem_share(struct gh_rm *rm, struct gh_rm_mem_parcel *mem_parcel)
 {
 	struct qcom_scm_vmperm *new_perms;
-	u64 src, src_cpy;
+	u64 src[2] = {0};
+    u64 src_cpy;
 	int ret = 0, i, n;
 	u16 vmid;
 
@@ -38,13 +39,13 @@ static int qcom_scm_gh_rm_pre_mem_share(struct gh_rm *rm, struct gh_rm_mem_parce
 			new_perms[n].perm |= QCOM_SCM_PERM_READ;
 	}
 
-	src = BIT_ULL(QCOM_SCM_VMID_HLOS);
+	src[0] = BIT_ULL(QCOM_SCM_VMID_HLOS);
 
 	for (i = 0; i < mem_parcel->n_mem_entries; i++) {
 		src_cpy = src;
 		ret = qcom_scm_assign_mem(le64_to_cpu(mem_parcel->mem_entries[i].phys_addr),
 						le64_to_cpu(mem_parcel->mem_entries[i].size),
-						&src_cpy, new_perms, mem_parcel->n_acl_entries);
+						&src_cpy[0], new_perms, mem_parcel->n_acl_entries);
 		if (ret)
 			break;
 	}
@@ -52,13 +53,13 @@ static int qcom_scm_gh_rm_pre_mem_share(struct gh_rm *rm, struct gh_rm_mem_parce
 	if (!ret)
 		goto out;
 
-	src = 0;
+	src[0] = 0;
 	for (n = 0; n < mem_parcel->n_acl_entries; n++) {
 		vmid = le16_to_cpu(mem_parcel->acl_entries[n].vmid);
 		if (vmid <= QCOM_SCM_MAX_MANAGED_VMID)
-			src |= BIT_ULL(vmid);
+			src[0] |= BIT_ULL(vmid);
 		else
-			src |= BIT_ULL(QCOM_SCM_RM_MANAGED_VMID);
+			src[0] |= BIT_ULL(QCOM_SCM_RM_MANAGED_VMID);
 	}
 
 	new_perms[0].vmid = QCOM_SCM_VMID_HLOS;
@@ -68,7 +69,7 @@ static int qcom_scm_gh_rm_pre_mem_share(struct gh_rm *rm, struct gh_rm_mem_parce
 		WARN_ON_ONCE(qcom_scm_assign_mem(
 				le64_to_cpu(mem_parcel->mem_entries[i].phys_addr),
 				le64_to_cpu(mem_parcel->mem_entries[i].size),
-				&src_cpy, new_perms, 1));
+				&src_cpy[0], new_perms, 1));
 	}
 
 out:
@@ -79,7 +80,7 @@ static int qcom_scm_gh_rm_pre_mem_share(struct gh_rm *rm, struct gh_rm_mem_parce
 static int qcom_scm_gh_rm_post_mem_reclaim(struct gh_rm *rm, struct gh_rm_mem_parcel *mem_parcel)
 {
 	struct qcom_scm_vmperm new_perms;
-	u64 src = 0, src_cpy;
+	u64 src[2] = {0}, src_cpy[2];
 	int ret = 0, i, n;
 	u16 vmid;
 
@@ -89,16 +90,16 @@ static int qcom_scm_gh_rm_post_mem_reclaim(struct gh_rm *rm, struct gh_rm_mem_pa
 	for (n = 0; n < mem_parcel->n_acl_entries; n++) {
 		vmid = le16_to_cpu(mem_parcel->acl_entries[n].vmid);
 		if (vmid <= QCOM_SCM_MAX_MANAGED_VMID)
-			src |= (1ull << vmid);
+			src[0] |= (1ull << vmid);
 		else
-			src |= (1ull << QCOM_SCM_RM_MANAGED_VMID);
+			src[0] |= (1ull << QCOM_SCM_RM_MANAGED_VMID);
 	}
 
 	for (i = 0; i < mem_parcel->n_mem_entries; i++) {
 		src_cpy = src;
 		ret = qcom_scm_assign_mem(le64_to_cpu(mem_parcel->mem_entries[i].phys_addr),
 						le64_to_cpu(mem_parcel->mem_entries[i].size),
-						&src_cpy, &new_perms, 1);
+						&src_cpy[0], &new_perms, 1);
 		WARN_ON_ONCE(ret);
 	}
 
diff --git a/include/linux/firmware/qcom/qcom_scm.h b/include/linux/firmware/qcom/qcom_scm.h
index c181671e505e..8ea2b0d9eec8 100644
--- a/include/linux/firmware/qcom/qcom_scm.h
+++ b/include/linux/firmware/qcom/qcom_scm.h
@@ -75,6 +75,8 @@ struct qcom_scm_mem_map_info {
 #define QCOM_SCM_PERM_EXEC       0x1
 #define QCOM_SCM_PERM_RW (QCOM_SCM_PERM_READ | QCOM_SCM_PERM_WRITE)
 #define QCOM_SCM_PERM_RWX (QCOM_SCM_PERM_RW | QCOM_SCM_PERM_EXEC)
+#define QCOM_SCM_VM_WORD_SIZE       64
+#define QCOM_SCM_MAX_VMID       128
 
 static inline void qcom_scm_populate_vmperm_info(
 		struct qcom_scm_current_perm_info *destvm, int vmid, int perm)
@@ -102,6 +104,7 @@ static inline void qcom_scm_populate_mem_map_info(
 extern bool qcom_scm_is_available(void);
 
 extern int qcom_scm_set_cold_boot_addr(void *entry);
+extern void qcom_scm_set_vmid_by_word(u64 *srcvm, int vmid);
 extern int qcom_scm_set_warm_boot_addr(void *entry);
 extern void qcom_scm_cpu_power_down(u32 flags);
 extern int qcom_scm_set_remote_state(u32 state, u32 id);
diff --git a/sound/soc/qcom/qdsp6/msm_audio_mem.c b/sound/soc/qcom/qdsp6/msm_audio_mem.c
index 0adab7bee3de..b37db86c1557 100644
--- a/sound/soc/qcom/qdsp6/msm_audio_mem.c
+++ b/sound/soc/qcom/qdsp6/msm_audio_mem.c
@@ -562,13 +562,14 @@ static int msm_audio_mem_free(struct dma_buf *dma_buf, struct msm_audio_mem_priv
 static int msm_audio_hyp_unassign(struct msm_audio_fd_data *msm_audio_fd_data)
 {
 	int ret = 0;
-	u64 src_vmid_unmap_list = BIT(VMID_LPASS) | BIT(VMID_ADSP_HEAP);
+	u64 src_vmid_unmap_list[2] = {0};
+	src_vmid_unmap_list[0] = BIT(VMID_LPASS) | BIT(VMID_ADSP_HEAP);
 	struct qcom_scm_vmperm dst_vmids_unmap[] = {{QCOM_SCM_VMID_HLOS,
 		PERM_READ | PERM_WRITE | PERM_EXEC}};
 
 	if (msm_audio_fd_data->hyp_assign) {
 		ret = qcom_scm_assign_mem(msm_audio_fd_data->paddr, msm_audio_fd_data->plen,
-				&src_vmid_unmap_list, dst_vmids_unmap, ARRAY_SIZE(dst_vmids_unmap));
+				&src_vmid_unmap_list[0], dst_vmids_unmap, ARRAY_SIZE(dst_vmids_unmap));
 		if (ret < 0) {
 			pr_err("%s: qcom assign unmap failed result = %d addr = 0x%llx size = %zu\n",
 				__func__, ret, msm_audio_fd_data->paddr, msm_audio_fd_data->plen);
@@ -647,13 +648,15 @@ static long msm_audio_mem_ioctl(struct file *file, unsigned int ioctl_num,
 	struct msm_audio_fd_data *msm_audio_fd_data = NULL;
 	struct msm_audio_mem_private *mem_data =
 			container_of(file->f_inode->i_cdev, struct msm_audio_mem_private, cdev);
-	u64 src_vmid_map_list = BIT(QCOM_SCM_VMID_HLOS);
+	u64 src_vmid_map_list[2] =  {0};
+	u64 src_vmid_unmap_list[2] =  {0};
 	struct qcom_scm_vmperm dst_vmids_map[] = {{VMID_LPASS, PERM_READ | PERM_WRITE},
 						{VMID_ADSP_HEAP, PERM_READ | PERM_WRITE}};
-	u64 src_vmid_unmap_list = BIT(VMID_LPASS) | BIT(VMID_ADSP_HEAP);
 	struct qcom_scm_vmperm dst_vmids_unmap[] = {{QCOM_SCM_VMID_HLOS,
 					PERM_READ | PERM_WRITE | PERM_EXEC}};
 
+	src_vmid_map_list[0] = BIT(QCOM_SCM_VMID_HLOS);
+	src_vmid_unmap_list[0] = BIT(VMID_LPASS) | BIT(VMID_ADSP_HEAP);
 	switch (ioctl_num) {
 	case IOCTL_MAP_PHYS_ADDR:
 		iosys_vmap = kzalloc(sizeof(struct msm_audio_fd_data), GFP_KERNEL);
@@ -695,7 +698,7 @@ static long msm_audio_mem_ioctl(struct file *file, unsigned int ioctl_num,
 			pr_err("%s get phys addr failed %d\n", __func__, ret);
 			return ret;
 		}
-		ret = qcom_scm_assign_mem(paddr, pa_len, &src_vmid_map_list,
+		ret = qcom_scm_assign_mem(paddr, pa_len, &src_vmid_map_list[0],
 			dst_vmids_map, ARRAY_SIZE(dst_vmids_map));
 		if (ret < 0) {
 			pr_err("%s: qcom_assign failed result = %d addr = 0x%llx size = %lu\n",
@@ -711,7 +714,7 @@ static long msm_audio_mem_ioctl(struct file *file, unsigned int ioctl_num,
 			pr_err("%s get phys addr failed %d\n", __func__, ret);
 			return ret;
 		}
-		ret = qcom_scm_assign_mem(paddr, pa_len, &src_vmid_unmap_list,
+		ret = qcom_scm_assign_mem(paddr, pa_len, &src_vmid_unmap_list[0],
 			dst_vmids_unmap, ARRAY_SIZE(dst_vmids_unmap));
 		if (ret < 0) {
 			pr_err("%s: qcom scm unassign failed result = %d addr = 0x%llx size = %lu\n",
-- 
2.34.1

